browser/ai_chat/content_agent_tool_provider.cc

```diff
@@ -0,0 +1,183 @@
+#include "chrome/browser/ui/browser_window/public/browser_window_interface_iterator.h"
+#include "chrome/common/actor.mojom.h"
+#include "components/optimization_guide/content/browser/page_content_proto_provider.h"
+#include "components/tabs/public/tab_interface.h"
+#include "content/public/browser/web_contents.h"
+#include "ui/base/page_transition_types.h"
+
+namespace ai_chat {
+
+ContentAgentToolProvider::ContentAgentToolProvider(
+    Profile* profile,
+    raw_ptr<actor::ActorKeyedService> actor_service)
+    : actor_service_(actor_service), profile_(profile) {
+  // Only create page actor tools if this profile is allowed to.
+  if (actor_service_) {
+    // Each conversation can have a different actor service task,
+    // and operate on a different set of tabs.
+    // If we want to delay creation of the task, we'll need to perhaps
+    // intercept all tool use calls and create or choose which task to use
+    // at that time. Tool::UseTool will have to change to
+    // ToolProvider::UseTool, or similar.
+    task_id_ = actor_service_->CreateTask();
+  }
+  CreateTools();
+}
+
+ContentAgentToolProvider::~ContentAgentToolProvider() = default;
+
+std::vector<base::WeakPtr<Tool>> ContentAgentToolProvider::GetTools() {
+  // Note: We don't have the ability to filter tools based on conversation
+  // capability here. But for now we don't need to as we only create the content
+  // agent tools if we don't have the actor_service_ (which is only provided for
+  // agent profiles).
+  std::vector<base::WeakPtr<Tool>> tool_ptrs;
+  tool_ptrs.reserve(tools_.size());
+  for (const auto& tool : tools_) {
+    tool_ptrs.push_back(tool->GetWeakPtr());
+  }
+  return tool_ptrs;
+}
+
+void ContentAgentToolProvider::StopAllTasks() {
+  if (actor_service_ && !task_id_.is_null()) {
+    // `success` sets whether the task ends as state kFinished or kCancelled
+    actor_service_->StopTask(task_id_, true /* success */);
+  }
+}
+
+actor::TaskId ContentAgentToolProvider::GetTaskId() {
+  return task_id_;
+}
+
+void ContentAgentToolProvider::GetOrCreateTabHandleForTask(
+    base::OnceCallback<void(tabs::TabHandle)> callback) {
+  if (!task_tab_handle_.Get()) {
+    // Create a new tab because we are only allowed to act on
+    // certain URLs, e.g. NTP. Safer to start on a blank page
+    // whilst this feature is focused on AI-initiated tasks instead
+    // of acting on existing tabs.
+    NavigateParams params(profile_, GURL(url::kAboutBlankURL),
+                          ui::PAGE_TRANSITION_FROM_API);
+    params.disposition = WindowOpenDisposition::NEW_BACKGROUND_TAB;
+    Navigate(&params);
+    content::WebContents* new_contents = params.navigated_or_inserted_contents;
+
+    task_tab_handle_ =
+        tabs::TabInterface::GetFromContents(new_contents)->GetHandle();
+  }
+  actor_service_->GetTask(task_id_)->AddTab(
+      task_tab_handle_, base::BindOnce(&ContentAgentToolProvider::TabAdded,
```

Comment:
Is there a reason we don't just invoke the callback directly here? Is it so the callback gets dropped if this class is destroyed?